File: soln-tree-problems/maximum-level-sum-binary-tree.java
/*
https://www.geeksforgeeks.org/find-level-maximum-sum-binary-tree/
https://leetcode.com/problems/maximum-level-sum-of-a-binary-tree/
*/

import java.io.*;
import java.util.*;
import java.lang.*;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxLevelSum(TreeNode root) {
        if(root==null)
            return -1;
        
        int max_sum=Integer.MIN_VALUE;
        int max_level=0;
        int level=0;
        
        Queue<TreeNode> queue=new LinkedList<>();
        queue.add(root);
        while(!queue.isEmpty()){
            int size=queue.size();
            level++;
            
            List<Integer> list=new ArrayList<Integer>();
            int sum=0;
            for(int i=0;i<size;i++){
                TreeNode node=queue.poll();
                list.add(node.val);
                sum+=node.val;
                
                if(node.left!=null)
                    queue.add(node.left);
                if(node.right!=null)
                    queue.add(node.right);
            }     
            if(max_sum<sum){
                max_sum=sum;
                max_level=level;
            }           
        }        
        return max_level;
    }
}

File: C++/soln-tree-binary-problems/maximum-level-sum-binary-tree.cpp
/**
 @author Farheen Bano
  
 References-
 https://www.geeksforgeeks.org/find-level-maximum-sum-binary-tree/
 https://leetcode.com/problems/maximum-level-sum-of-a-binary-tree/
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxLevelSum(TreeNode* root) {
 
        if(root==NULL){
            return -1;
        }
        
        int max_sum=INT_MIN;
        int level=0;
        int cur_level=0;
            
        queue<TreeNode *> qu;
        qu.push(root);
        while(!qu.empty()){
            int size=qu.size();
            cur_level++;
            
            int cur_sum=0;            
            for(int i=0;i<size;i++){
                TreeNode* node=qu.front();
                qu.pop();
                cur_sum+=node->val;
                   
                if(node->left!=NULL)
                    qu.push(node->left);
                if(node->right!=NULL)
                    qu.push(node->right);
            }
            if(max_sum<cur_sum){
                max_sum=cur_sum;
                level=cur_level;
            }

        }
        return level;
    }
};